cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java

```diff
@@ -422,27 +413,77 @@ public static void jvmSystemProperties(List<EnvVar> envVars, JvmOptions jvmOptio
-        StringBuilder jvmPerformanceOpts = new StringBuilder();
-
-        Map<String, String> xx = jvmOptions != null ? jvmOptions.getXx() : null;
-        if (xx != null) {
-            xx.forEach((k, v) -> {
-                jvmPerformanceOpts.append(' ').append("-XX:");
-
-                if ("true".equalsIgnoreCase(v))   {
-                    jvmPerformanceOpts.append("+").append(k);
-                } else if ("false".equalsIgnoreCase(v)) {
-                    jvmPerformanceOpts.append("-").append(k);
-                } else  {
-                    jvmPerformanceOpts.append(k).append("=").append(v);
-                }
-            });
+        parseJvmPerformanceOptions(jvmOptions)
+            .map(envVar -> ContainerUtils.createEnvVar(AbstractModel.ENV_VAR_KAFKA_JVM_PERFORMANCE_OPTS, envVar))
+            .ifPresent(envVars::add);
+    }
+
+    /**
+     * Parses JVM performance options from {@link JvmOptions#getXx()} property into a space-separated string of JVM flags.
+     * If performance options object is null or XX options are not configured then {@link Optional#empty()} will be returned.
+     *
+     * @param jvmOptions JVM options to parse
+     * @return optional comma-separated string of JVM flags
+     */
+    private static Optional<String> parseJvmPerformanceOptions(JvmOptions jvmOptions) {
+        return Optional.ofNullable(jvmOptions)
+            .map(JvmOptions::getXx)
+            .map(Map::entrySet)
+            .map(entrySet -> entrySet.stream()
+                .sorted(ModelUtils::compareJvmPerformanceOption)
+                .map(ModelUtils::toJvmPerformanceFlag)
+                .collect(Collectors.joining(" ")))
+            .filter(envVar -> !envVar.isEmpty());
+    }
+
+    /**
+     * Compares two JVM option for sorting. The options are sorted by priority or if not prioritized by name.
+     * The prioritized options are placed at the beginning to ensure correctness of JVM flags.
+     * For example `-XX:+UnlockDiagnosticVMOptions` flag must be entered before `-XX:+PrintNMTStatistics`.
+     *
+     * @param jvmOption1 first JVM option to compare with key holding flag name placed in the key
+     * @param jvmOption2 second JVM option to compare with key holding flag name placed in the key
+     * @return comparison result [-1, 0, 1]
+     */
+    private static int compareJvmPerformanceOption(Map.Entry<String, String> jvmOption1, Map.Entry<String, String> jvmOption2) {
+        final var isJvmOption1Prioritized = isPriorityJvmPerformanceOption(jvmOption1);
+        final var isJvmOption2Prioritized = isPriorityJvmPerformanceOption(jvmOption2);
+        if (isJvmOption1Prioritized && isJvmOption2Prioritized) {
+            return 0;
+        }
+        if (isJvmOption1Prioritized) {
+            return -1;
+        }
+        if (isJvmOption2Prioritized) {
+            return 1;
         }
+        return jvmOption1.getKey().compareTo(jvmOption2.getKey());
```

Comment:
The alphabetical sorting is not really required, I just added it to produce more consistent results, but if you think it's better to try preserve original order (in-case for example `LinkedHashMap` was used) I can remove it.